### ISSUE TITLE
How to handle by admins the new skill suggestions

### ISSUE THREAD

**Issue by XaviP** [FreedomCoop/valuenetwork#143]
@bhaugen said: Now we need to figure out how to handle these new skill suggestions. Basically, a coop admin needs to review them and (maybe) turn them into new work resource types.
But they are also the start of a conversation with whoever suggested them, so they might need some comments, which could go on the worker's discussion page. And then maybe if the new skill suggestion is enacted, the worker could see a little "thanks for the suggestion" tag on their skills list on their profile?
Also, should coop admins be notified about these, like membership requests?
In other words, is the workflow here maybe a lot like membership requests? Or is that too much ceremony?

**Comment by bhaugen** [comment]
Maro writes in telegram,
> I have also a question about skills..I dont have clear how/where as admin I manage the suggestion of users about it...As user marito (in test of curse) I suggest training as a skill and as admin profile (maro) dont know where to find it and how to add it in order that it appears in the skills list...

**Comment by bhaugen** [comment]
If nobody has any better ideas, I'll handle these like membership requests:
* add a couple of fields to SkillSuggestion:
    * resource_type (when a new resource type is created for the SkillSuggestion)
    * state (similar to MembershipRequest state: "new", "accepted", maybe "rejected" instead of "declined"?
* Add an email notification to coop admins
* Add a page like the membership request table where coop admins can see skill suggestions, with a button to create a new resource type based on the suggestion.
* Add the suggested skill to the member who suggested it? Or not? What do you all think about that one?

Anything else?

Also @XaviP can we add migrations again? And should I add the new fields and migration in master and then do the rest of the work in a branch?

**Comment by bhaugen** [comment]
Hearing no objections, I started working on the plan above. I think it's probably better to copy an existing flow that the coop admins somewhat understand rather than design a new one.

But as of the proposal above, I am not adding the automatic todo/task. Do you all think that is useful, and want it? Or not so much?

**Comment by Holofractographic** [comment]
Hi ok for me is a good plan, if we can know who made the suggestion also we can communicate with him/her if the suggestion already exist as similar or to aks for more explanation or to tell him/her that the suggestion was accepted..

**Comment by XaviP** [comment]
> Also @XaviP can we add migrations again?

Yes, no prob.

> And should I add the new fields and migration in master and then do the rest of the work in a branch?

The problem is when branches in parallel working on migrations in the same app. It's not the case now, but if you have a definitive dessign of models (no changes later), maybe is a good idea to push migrations in master before to open new branch to work. As you prefer.

**Comment by bhaugen** [comment]
@Holofractographic 
> if we can know who made the suggestion also we can communicate with him/her if the suggestion already exist as similar or to aks for more explanation or to tell him/her that the suggestion was accepted..

I'm adding a link to the suggester's discussion page in the table of suggestions, so a coop admin can talk to them about it.

When the suggestion is accepted, they will also see a little note on their profile saying "thanks for the suggestion".

**Comment by bhaugen** [comment]
@XaviP 
> The problem is when branches in parallel working on migrations in the same app. It's not the case now, but if you have a definitive dessign of models (no changes later), maybe is a good idea to push migrations in master before to open new branch to work. As you prefer.

I already started a new branch, and am not sure the changes will be that definitive, so I'll try to move quickly to a working version and merge it in. Since the changes will not affect anything else, they should be fairly safe.

Then I think we need more discussion about skills in general. I know some people want them to have more structure. But that would be for a different issue.

**Comment by bhaugen** [comment]
I got the first part of this done:
* two new fields on SkillSuggestion, with a new work.migration
* a skill suggestions page for coop admins with buttons for both of the new fields, which works quite a lot like the membership and join request pages

And I merged it all into master locally, and pushed to github.

Next will be a notification for coop admins of the new skill suggestion, which I will do in my branch.

**Comment by bhaugen** [comment]
@XaviP @bum2 I got the notifications done in my branch. They seem a little unsafe to deploy to prod ocp without testing in testocp.  I have not yet pushed my branch to github, or merged it into master and pushed that. What do you think I should do?

(P.S. I also worked out a procedure for testing notifications locally in the shell, which I will post as a separate issue.)

**Comment by bhaugen** [comment]
Notifiction testing procedure: https://github.com/FreedomCoop/valuenetwork/issues/155

**Comment by bhaugen** [comment]
@fosterlynn says I should push my branch, so I did:
https://github.com/FreedomCoop/valuenetwork/tree/skill_workflow

**Comment by XaviP** [comment]
If you merge to master we can deploy in testocp and test notifications.

**Comment by bhaugen** [comment]
Merged and pushed. Let me know when it's in testocp and I'll try a skill suggestion and so a notification.

**Comment by bum2** [comment]
Deployed in testocp.
In local it work good! just missing the link in organization page sub-nav. 
Also missing the 'comments' page link in the same sub-nav, but it shows either membership requests comments and join requests comments.... will be good to only show membership comments in that page and let the join-requests only to project coordinators.

**Comment by bum2** [comment]
it throws an error when pushing the create skill button in testocp:
`BdbQuit at /accounting/create-skill/2/
No exception message supplied`
when running the line 13254 `if form.is_valid(): `

**Comment by bhaugen** [comment]
OOps! Stupid programmer error, left a pdb halt in the code. Commented out, committed and pushed to master.

**Comment by bhaugen** [comment]
I would like to add one more minor feature to skill suggestions: on the member's profile, when their skill suggestion is accepted and a new resource type is created for it, to have a little note on their profile alongside the new skill saying "thanks for the suggestion". I'll do that some time today.

**Comment by bhaugen** [comment]
@XaviP @bum2 please let me know when my last fix is deployed to testocp. I want to test it.

**Comment by XaviP** [comment]
Done

**Comment by bhaugen** [comment]
I got an email notification about my skill suggestion, but here's the subject line:
``` [] New skill suggestion ```
Whereas here's the subject line of a comment notification from testocp:
``` [FreedomCoop OCP] New comment in membership request ```

I think this is the EMAIL_SUBJECT_PREFIX in local_settings.py. I wonder why it did not appear on the skill suggestion notice. Is it something I did differently? I can't see any difference in the short.txt templates.

**Comment by bhaugen** [comment]
Followup issue: https://github.com/FreedomCoop/valuenetwork/issues/156

**Comment by bhaugen** [comment]
Added a little "thanks for the suggestion" label to profile skills when you suggested the skill and it became a new resource type and comes back at you.

With that, unless I need to do something about the notice email subject line, I think this issue is ready to close, and the branch deleted.

**Comment by bum2** [comment]
Ok, updated repo in testocp, but now we have no parent field... 
I think it makes no sense to list all skills without any kind of grouping or hierachy... i'm used to do this kind of thing as a tree, normally with a drag'n'drop jquery to manage the MPTT tree in the admin side, and a chooser ready for mptt data (also jquery plugins) in the user side, to choose skills from the tree.
There's a django-mptt plugin to help manage this kind of data (i've tried long ago) but nowadays with just a 'parent' field we can build the structure.
I will search for the jquery plugins i used to manage the trees, and refresh my memory about mptt fields in django. (MPTT: Modified Preorder Tree Traversal, is a super fast way to read complex trees in SQL, it don't use recursivity).

**Comment by bhaugen** [comment]
I commented out the parent field because it was confusing people. We can put it back by removing the comments, when we decide how to structure types of work.

**Comment by fosterlynn** [comment]
@bum2 I would recommend we see if the faceted classification system we currently have will work for skills.  It can give you at least a 2 level pseudo-heirarchical structure if you like.  Let's work on it together.  Or if you give me a nice long list of skills, I will put together a sample set of facets and facet values and we can see if it will work.

We could work in testocp if you like.  If you put in a bunch of skills as resource types, I will try some classification.  Or send me a list, I will put them in, whichever is easiest.

If it doesn't work, then we can add a more hierarchical classification sytem.
